### PR TITLE
Make search and building of index parallel

### PR DESCRIPTION
**Pull request by thaqibm**
Make search and building of index parallel

### CODE REVIEW

File: src/benchmark.sh [hunk 1]
```diff
@@ -26,8 +26,8 @@ else
 fi
 
 # Run the faiss benchmarking
-FAISS_FILE="${SCRIPT_DIR}/faiss_run.py"
-python3 "$FAISS_FILE" --input-vec "$WIKIDATA_FILE" --data-dir "$DATA_DIR"
+# FAISS_FILE="${SCRIPT_DIR}/faiss_run.py"
+# python3 "$FAISS_FILE" --input-vec "$WIKIDATA_FILE" --data-dir "$DATA_DIR"
```
Comment (nikhilgarg28) [inline src/benchmark.sh]: Why are you commenting these out?

File: src/search.rs [hunk 2]
```diff
@@ -135,7 +137,7 @@ impl<const N: usize> ANNIndex<N> {
         // Trees hold an index into the [unique_vecs] list which is not
         // necessarily its id, if duplicates existed
         let all_indexes: Vec<usize> = (0..unique_vecs.len()).collect();
-        let trees: Vec<_> = (0..num_trees)
+        let trees: Vec<_> = (0..num_trees).into_par_iter()
             .map(|_| Self::build_a_tree(max_size, &all_indexes, &unique_vecs))
```
Comment (nikhilgarg28) [inline src/search.rs]: Is this rustfmt compliant?

File: src/search.rs [hunk 3]
```diff
@@ -182,10 +184,11 @@ impl<const N: usize> ANNIndex<N> {
...
     ) -> Vec<(i32, f32)> {
-        let mut candidates = HashSet::new();
-        for tree in self.trees.iter() {
-            Self::tree_result(query, top_k, tree, &mut candidates);
-        }
+        let candidates = DashSet::new();
+        self.trees.par_iter()
+            .for_each(|tree| {
+                Self::tree_result(query, top_k, tree, &candidates);
+            });
```
Comment (nikhilgarg28) [inline src/search.rs]: Is this rustfmt compliant?

**Review by nikhilgarg28 — CHANGES_REQUESTED** [review]
Could you please report how parallelizing affects build and query times against the benchmark?

Also requesting changes on some other things mentioned below.

**Comment by thaqibm** [comment]
Ran the benchmarks on Macbook air M1 (8gb ram, 8 cores).

## Build Tree for max_node_size=30
image

image

Build tree has a 2-3x speedup

## bulk search for max_node_size=30
image

image

Bulk search gets a ~1.5x speedup.

other max_node_sizes have similar speedup. 

[baseline benchmark](https://github.com/fennel-ai/fann/files/11747471/baseline.txt)
[parallel benchmark](https://github.com/fennel-ai/fann/files/11747472/par.txt)

**Review by nikhilgarg28 — APPROVED** [review]

**Pull request merged**